selection_tools/src/HistogramFiller.cpp

```diff
@@ -219,31 +313,58 @@ void HistogramFiller::fill_first_muon_from_alp_selection_hists(const Particle* p
   histSets["alp_selection_os_minlxy-muon"]->fill(particle, event);
 };
 
+void HistogramFiller::fill_2d_hists(const Particle* particle_1, const Particle* particle_2, string sign)
+{
+  if(!particle_1 || !particle_2) return;
+  histSets[sign+"_dimuon"]->fill_2d(particle_1, particle_2);
+}
+
 void HistogramFiller::save_histograms(std::string output_path)
 {
   auto output_file = new TFile(output_path.c_str(), "recreate");
   output_file->cd();
   output_file->mkdir("final_selection");
-  output_file->mkdir("intermediate_selections");
   output_file->mkdir("alp_selections");
-  
-  for(auto &[hist_name, hist_set] : histSets){
-    if(hist_name.substr(0,15) == "final_selection"){
-      output_file->cd("final_selection");
-    }
-    if(hist_name.substr(0,3) == "sel"){
-      output_file->cd("intermediate_selections");
-    }
-    if(hist_name.substr(0,5) == "alp_s"){
-      output_file->cd("alp_selections");
-    }
-    else if(hist_name.substr(0,3) == "alp"){
-      output_file->cd("alp");
+
+  if (reduce_hists_){
+    for(auto &[hist_name, hist_set] : histSets){
+      if(hist_name.substr(0,15) == "final_selection"){
+        output_file->cd("final_selection");
+      }
+      if(hist_name.substr(0,5) == "alp_s"){
+        output_file->cd("alp_selections");
+      }
+      for(auto &[tmp, hist] : hist_set->hists){
+        hist->Write();
+      }
+      output_file->cd();
     }
-    for(auto &[tmp_2, hist] : hist_set->hists){
-      hist->Write();
+  }
+  else {
+    output_file->mkdir("intermediate_selections");
+    output_file->mkdir("alp");
+    
+    for(auto &[hist_name, hist_set] : histSets){
+      if(hist_name.substr(0,15) == "final_selection"){
+        output_file->cd("final_selection");
+      }
+      if(hist_name.substr(0,3) == "sel"){
+        output_file->cd("intermediate_selections");
+      }
+      if(hist_name.substr(0,5) == "alp_s"){
+        output_file->cd("alp_selections");
+      }
+      else if(hist_name.substr(0,3) == "alp"){
+        output_file->cd("alp");
+      }
+      for(auto &[tmp, hist] : hist_set->hists){
+        hist->Write();
+      }
+      for(auto &[tmp, hist2d] : hist_set->hists2d){
+        hist2d->Write();
+      }
+      output_file->cd();
```

Comment:
I have a feeling this could be simplified. Maybe it should be one for-loop with `if(reduce_hists) continue` a some point?